python/ray/autoscaler/autoscaler.py

```diff
@@ -205,11 +205,21 @@ def _update(self):
                     self.resource_demand_scheduler.get_nodes_to_launch(
                         self.provider.non_terminated_nodes(tag_filters={}),
                         self.pending_launches.breakdown(),
-                        resource_demand_vector))
+                        resource_demand_vector,
+                        self.load_metrics.get_static_resource_usage()))
                 # TODO(ekl) also enforce max launch concurrency here?
                 for node_type, count in to_launch:
                     self.launch_new_node(count, node_type=node_type)
 
+        # TODO (Alex): There's a race condition here where `num_pending` is
+        # derived from `pending_launches` and `len(nodes)` comes from
+        # `self.workers` and we have no synchronization to update them at the
+        # same time. This line ensures that we don't underestimate the number
+        # of workers. This may increase the latency of worker startup by an
+        # update (or so) which is better than a correctness issue around
+        # `max_workers`.
+        nodes = self.workers()
```

Comment:
Should this be inside the `if self.resource_demand_scheduler` block?

Reply:
I think this is the right place (maybe the whitespacing wasn't clear). This is used for the `num_workers` calculation below, and the race is with the `node_launcher` threads.

Reply:
Isn't the issue that we don't re-read `self.pending_launches.value` after launching more nodes in the block above? Though it does seem there is a race condition where we might miss some recently enqueued node requests if no threads have picked them up.

I would propose instead adding `nodes = self.workers; num_pending = self.pending_launches.value` to the end of that if block at the end instead, to recompute both in this case.

Reply:
hmm the race condition i was thinking of was the case in which there is a queue (perhaps from the previous run of `_update`), then we get the current worker nodes, a new node launches, then we get the number of pending nodes (so we lose a node).

i think what you're describing is a completely deterministic bug that can be solved by adding refreshing `nodes` inside the resource demand scheduler block

Reply:
Yeah so we need to refresh both right? Why not do this:

- remove this line
- add to the block above
```
nodes = self.workers
num_pending = self.pending_launches.value
```

Reply:
I'll move this if you want, but I don't think we're talking about the same bug

Reply:
I think this fixes what you're talking about too. If the bug is something related to the existing code, we can maybe address that in a separate PR.